File: tests/cases/py_call_service/module.py
import yaml
import pprint

from skull.txn import *
from skull.txndata import *
from skull import logger as Logger

from common import metrics as Metrics
from common.proto import *

def module_init(config):
    print("py module init")
    print("config: ", end=' ')
    pprint.pprint(config)

    Logger.trace('py module init: trace test')
    Logger.debug('py module init: debug test')
    Logger.info('1', 'py module init: info test')
    Logger.warn('2', 'py module init: warn test', 'no suggestion')
    Logger.error('3', 'py module init: error test', 'no solution')
    Logger.fatal('4', 'py module init: fatal test', 'no solution')
    return True

def module_release():
    print("py module release")
    return

def module_unpack(txn, data):
    print("data: %s" % data, end=' ')
    Logger.info('5', 'receive data: {}'.format(data.decode('UTF-8')))
    example_msg = txn.data()
    example_msg.data = data

    return len(data)

def module_pack(txn, txndata):
    mod_metrics = Metrics.module()
    mod_metrics.response.inc(1)

    mod_dymetrics = Metrics.transaction('test')
    mod_dymetrics.response.inc(1)

    if txn.status() != Txn.TXN_OK:
        txndata.append('error')
        Logger.error('6', 'module_pack error', 'no solution')
    else:
        example_msg = txn.data()
        print("pack data: %s" % example_msg.data)
        Logger.info('7', 'module_pack: data sz: {}'.format(len(example_msg.data)))

        txndata.append(example_msg.data)

def module_run(txn):
    print("module_run")
    mod_metrics = Metrics.module()
    mod_metrics.request.inc(1)

    mod_dymetrics = Metrics.transaction('test')
    mod_dymetrics.request.inc(1)

    # Assemble api request
    get_req_msg = service_s1_get_req_pto.get_req()
    get_req_msg.name = "hello"

    # invoke iocall to s1 service
    ret = txn.iocall('s1', 'get', get_req_msg, 0, _api_cb)

    print("iocall ret: {}".format(ret))
    return True

def _api_cb(txn, iostatus, api_name, request_msg, response_msg):
    print("api_cb: iostatus: {}, api_name: {}, request_msg: {}, response_msg: {}".format(
            iostatus, api_name, request_msg, response_msg))

    example_msg = txn.data()
    example_msg.data = str(response_msg.response).encode('UTF-8')
    return True

File: tests/cases/call_multi_services/module.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <string>
#include <iostream>

#include "skullcpp/api.h"
#include "skull_metrics.h"
#include "skull_protos.h"
#include "config.h"

static skullcpp::metrics::module moduleMetrics;

static
void module_init(const skull_config_t* config)
{
    std::cout << "module(test): init" << std::endl;
    SKULL_LOG_TRACE("skull trace log test %d", 1);
    SKULL_LOG_DEBUG("skull debug log test %d", 2);
    SKULL_LOG_INFO("1", "skull info log test %d", 3);
    SKULL_LOG_WARN("1", "skull warn log test %d", 4, "ignore, this is test");
    SKULL_LOG_ERROR("1", "skull error log test %d", 5, "ignore, this is test");
    SKULL_LOG_FATAL("1", "skull fatal log test %d", 6, "ignore, this is test");

    // Load skull_config to skullcpp::Config
    skullcpp::Config::instance().load(config);
}

static
void module_release()
{
    std::cout << "module(test): released" << std::endl;
    SKULL_LOG_INFO("5", "module(test): released");
}

static
size_t module_unpack(skullcpp::Txn& txn, const void* data, size_t data_sz)
{
    moduleMetrics.request.inc(1);

    std::cout << "module_unpack(test): data sz: " << data_sz << std::endl;
    SKULL_LOG_INFO("2", "module_unpack(test): data sz:%zu", data_sz);

    // deserialize data to transcation data
    auto& example = (skull::workflow::example&)txn.data();
    example.set_data(data, data_sz);
    return data_sz;
}

static
int svc_api_callback(skullcpp::Txn& txn, skullcpp::Txn::IOStatus status,
                     const std::string& apiName,
                     const google::protobuf::Message& request,
                     const google::protobuf::Message& response)
{
    const auto& apiReq  = (skull::service::s1::get_req&)request;
    const auto& apiResp = (skull::service::s1::get_resp&)response;

    std::cout << "svc_api_callback.... apiName: " << apiName << std::endl;
    std::cout << "svc_api_callback.... request: " << apiReq.name() << std::endl;
    std::cout << "svc_api_callback.... response: " << apiResp.response() << std::endl;

    auto& example = (skull::workflow::example&)txn.data();
    example.set_data(example.data() + ", " + apiResp.response());
    return 0;
}

static
int module_run(skullcpp::Txn& txn)
{
    auto& example = (skull::workflow::example&)txn.data();

    std::cout << "receive data: " << example.data() << std::endl;
    SKULL_LOG_INFO("3", "receive data: %s", example.data().c_str());

    // Call service
    skull::service::s1::get_req req;
    req.set_name(example.data());

    skullcpp::Txn::IOStatus ret =
        txn.iocall("s1", "get", req, 0, svc_api_callback);

    std::cout << "ServiceCall ret: " << ret << std::endl;

    // Call service again
    skullcpp::Txn::IOStatus ret1 =
        txn.iocall("s1", "get", req, 0, svc_api_callback);

    std::cout << "ServiceCall ret1: " << ret1 << std::endl;
    return 0;
}

static
void module_pack(skullcpp::Txn& txn, skullcpp::TxnData& txndata)
{
    moduleMetrics.response.inc(1);
    auto& example = (skull::workflow::example&)txn.data();

    std::cout << "module_pack(test): data sz: " << example.data().length() << std::endl;
    SKULL_LOG_INFO("4", "module_pack(test): data sz:%zu", example.data().length());
    txndata.append(example.data().c_str(), example.data().length());
}

/******************************** Register Module *****************************/
static skullcpp::ModuleEntry module_entry = {
    module_init,
    module_release,
    module_run,
    module_unpack,
    module_pack
};

SKULLCPP_MODULE_REGISTER(&module_entry)
